src/flamegpu/simulation/detail/AbstractSimRunner.cu

```diff
@@ -0,0 +1,114 @@
+      , total_runners(_total_runners)
+      , verbosity(_verbosity)
+      , err_ct(_err_ct)
+      , next_run(_next_run)
+      , plans(_plans)
+      , step_log_config(std::move(_step_log_config))
+      , exit_log_config(std::move(_exit_log_config))
+      , run_logs(_run_logs)
+      , log_export_queue(_log_export_queue)
+      , log_export_queue_mutex(_log_export_queue_mutex)
+      , log_export_queue_cdn(_log_export_queue_cdn)
+      , err_detail(_err_detail)
+      , isSWIG(_isSWIG) {
+}
+void AbstractSimRunner::start() {
+    this->thread = std::thread(&AbstractSimRunner::main, this);
+    // Attempt to name the thread
+#ifdef _MSC_VER
+    std::wstringstream thread_name;
+    thread_name << L"CUDASim D" << device_id << L"T" << runner_id;
+    // HRESULT hr =
+    SetThreadDescription(this->thread.native_handle(), thread_name.str().c_str());
+    // if (FAILED(hr)) {
+    //     fprintf(stderr, "Failed to name thread 'CUDASim D%dT%u'\n", device_id, runner_id);
+    // }
+#else
+    std::stringstream thread_name;
+    thread_name << "CUDASim D" << device_id << "T" << runner_id;
+    // int hr =
+    pthread_setname_np(this->thread.native_handle(), thread_name.str().c_str());
+    // if (hr) {
+    //     fprintf(stderr, "Failed to name thread 'CUDASim D%dT%u'\n", device_id, runner_id);
+    // }
+#endif
+}
+void AbstractSimRunner::join() {
+    if (this->thread.joinable()) {
+        this->thread.join();
+    }
+}
+
+void AbstractSimRunner::runSimulation(int plan_id) {
+    // Update environment (this might be worth moving into CUDASimulation)
+    auto &prop_map = model->environment->properties;
+    for (auto &ovrd : plans[plan_id].property_overrides) {
+        auto &prop = prop_map.at(ovrd.first);
+        memcpy(prop.data.ptr, ovrd.second.ptr, prop.data.length);
+    }
+    // Set simulation device
+    std::unique_ptr<CUDASimulation> simulation = std::unique_ptr<CUDASimulation>(new CUDASimulation(model, isSWIG));
+    // Copy steps and seed from runplan
+    simulation->SimulationConfig().steps = plans[plan_id].getSteps();
+    simulation->SimulationConfig().random_seed = plans[plan_id].getRandomSimulationSeed();
+    simulation->SimulationConfig().verbosity = Verbosity::Default;
+    if (verbosity == Verbosity::Quiet)  // Use quiet verbosity for sims if set in ensemble but never verbose
+        simulation->SimulationConfig().verbosity = Verbosity::Quiet;
+    simulation->SimulationConfig().telemetry = false;   // Never any telemtry for individual runs inside an ensemble
+    simulation->SimulationConfig().timing = false;
+    simulation->CUDAConfig().device_id = this->device_id;
+    simulation->CUDAConfig().is_ensemble = true;
+    simulation->CUDAConfig().ensemble_run_id = plan_id;
+    simulation->applyConfig();
+    // Set the step config directly, to bypass validation
+    simulation->step_log_config = step_log_config;
+    simulation->exit_log_config = exit_log_config;
+    // TODO Set population?
```

Comment:
Is this TODO still relevant?

Reply:
Todo is taken from code originally in `CUDAEnsemble`, not required as we opt'd that ensembles must have population created via init function. Will remove/change